### ISSUE TITLE
[Proposal][WIP] Auth and Policy Design

### ISSUE THREAD

**Issue by jjcollinge** [dapr/dapr#1617]
# Auth and Policy Design
> Caveat this is very much a WIP brain dump and will be revised and cleaned up. Please feel free to make suggestions and fixes!

> This proposal is not concerned with authenticating a user app to the Dapr's API which is already covered by https://github.com/dapr/dapr/pull/1606.

## Contents
- [Dapr Auth & Policy Review](#dapr-auth-&-policy-review)
- [Proposal 1: Policy as a Building Block](#proposal-1-policy-as-a-building-block)
- [Proposal 2: Auth & Policy Middleware](#proposal-2-auth-&-policy-middleware)

## Definitions
- **Authentication** is concerned with who a caller is (identity).
- **Authorization** is concerned with what a caller can do (permissions). Authorization is typically done after a successful authentication.
- **Policy** is concerned with applying a set of defined rules to data to check whether it satisfies said rules. In this case, policy is considered a generalization of authorization.
- **Policy Decision** is the deciding whether data satisfies a policy
- **Policy Enforcement** the action taken on a Policy Decision

# Dapr Auth & Policy Review
Firstly, I have reviewed the existing middleware design and the authentication/authorization mechanisms provided by Dapr. Dapr doesn't currently have a concept of policy.

### [mTLS vs JWTs](https://github.com/dapr/docs/blob/master/concepts/security/README.md#network-security)
Dapr already supports mTLS for service to service calls using [Sentry](https://github.com/dapr/dapr/tree/master/cmd/sentry) as a CA. mTLS means that the connection between the two services is encrypted and therefore protected from mitm attacks. Additionally, it means that both sides of the connection have verified identities which can be used to authorize the connection (Service A can talk to Service B but not Service C). However, there is a higher level concept that can be applied on top of this network security which includes who the service is making the request on behalf of, what is their identity and what authorization do they have. For instance, Service A might be calling Service B on behalf of User X and that user may have a role that allows them access to additional data compared to whether User Y was making the initial call. A typical method of sharing the identity of the user is to use JSON Web Tokens (JWTs). The JWTs acts as a verified credential and contains claims that represent additional properties the user pocesses i.e. role, groups, etc. Please refer to [Middleware Pipeline Design](#middleware-pipeline-design) and [Bearer HTTP Middleware](#bearer-http-middleware) for more details on why Dapr doesn't currently fully support this model.
- [ ] Consider integrating Sentry with SPIFFE/SPIRE?
- [ ] Should we provide finer grained mTLS controls on which identities a Dapr service trusts?

### Middleware Pipeline Design
![middleware](https://user-images.githubusercontent.com/7241190/83652368-7261df80-a5b2-11ea-99ea-6f97f0063f78.png)

Middleware is a core extensibilty capability for Dapr's HTTP and gRPC API. Having a robust and configurable middleware story for Dapr is very important for production usage.

Dapr's current **middleware** design doesn't lend itself to authenticating calls from other Dapr services (service invocation) as the middleware is only applied on the local side of the requesting app. In the above diagram this would mean you'd need to express the API authentication middleware for App B in App A's pipeline config (and any other clients). See [Proposal 2](#proposal-2-auth-&-policy-middleware) for how we may modify this.
The current design does lend itself to authenticating calls to the Dapr API from the local app. Therefore, the app can generate or proxy a user JWT and send it with the request to the Dapr API, an authentication middleware can then use this to determine whether the request should be allowed. This is different to authenticating the user app itself to the Dapr API which is done via a shared API Token.

### [Bearer HTTP Middleware](https://github.com/dapr/components-contrib/blob/master/middleware/http/bearer/bearer_middleware.go) (Non-interactive)
This middleware extracts a bearer token from the HTTP request's `Authorization` header and then verifies it. The bearer token is formatted as a JSON Web Token (JWT). The middleware uses an `issuerURL` to fetch the JSON Web Key Sets (JWKS) from the Authorization Server. The `IssuerURL` could be an OIDC provider (which supports [Content Discovery](https://swagger.io/docs/specification/authentication/openid-connect-discovery/) or a vanilla JWKS URI (`https://YOUR_DOMAIN/.well-known/jwks.json`) that returns the JWKS directly. We currently expect an OIDC provider and do not handle this fallback condition. The JWT is then decoded and verified against this information to ensure the token is authentic (signed) and valid (not expired and correct information). As mentioned in [mTLS vs JWTs](#mtls-vs-jwts), JWTs are valuable to tell a service who they are calling on behalf of (or for service accounts). Verfiying the JWT allows us to trust the credential is authentic and authenticate the call. However, JWTs are often used to add additional claims about what that user has permission to do. We currently do not parse and verify any claims from the JWT which means we cannot provide any authorization on top of the authentication and the author has to do it in their app.
There are also other methods of presenting an auth token beyond JWTs that we may need to consider in the future depending on the clients we want to support but this isn't a priority.
- [ ] [Parse claims and provide mechanism to validate values (Authorization)](https://github.com/dapr/components-contrib/issues/189)
- [ ] Document expected use case of JWTs
- [ ] Add support for raw JWKS URIs (low priority)

### [OAuth2 HTTP Middleware](https://github.com/dapr/components-contrib/blob/master/middleware/http/oauth2/oauth2_middleware.go) (Interactive)
The existing OAuth2 middleware allows you to protect an API by performing an OAuth2 [Authorization Code Flow](https://www.oauth.com/oauth2-servers/pkce/authorization-code-exchange/) to obtain an access token from an Authorization provider. This is an interactive flow typically used on web, mobile, etc. for a user logging in and out of an application. The access token is stored in a session (server-side) for future requests and added to a request header so it can be accessed by upstream handlers (including the user's handler). The Authorization server may also return `id_tokens` which is typically the case for a single page application (SPA) accessing APIs directly without a "backend" component ([Implicit Grant Flow](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-implicit-grant-flow)). The `id_token` is commonly used in the client, where as the `access_token` is for the API resource access. Therefore, the `id_token` is currently discarded in the OAuth2 middleware. If the user code wishes to retrieve user information, they can use the access token to request it from a user endpoint on the Authorization server. Session state is currently not shared across instances of sidecars (i.e. 3 instance of the same app each with their own sidecar), therefore you could be prompted to login multiple times if your request is distributed to a new instance without the existing session state. Things to consider here are sticky sessions (or redirects) or distributed session state. In general, I'm not currently clear on the expected use case for needing this middleware for Dapr as a sidecar. Dapr also doesn't currently have middleware support for the [Client Credential Flow](https://auth0.com/docs/flows/concepts/client-credentials) which is a non interactive OAuth2 flow which is more widely used with APIs.
- [ ] Validate the use case for this middleware - do we need it? What is the expected use case with dapr?
- [ ] Address comments https://github.com/dapr/components-contrib/issues/155#issuecomment-575066554
- [ ] Should we default the `authHeaderName` to `Authorization`?
- [ ] Should we support the Client Credential Flow?

### HTTP Middleware Selectors
Typically you want to ONLY protect certain routes on your API and leave others open. Dapr doesn't currently support using selectors to apply middleware to specific routes. This PR #1556 would enable this.
- [ ] Review and merge #1556

### gRPC Middleware
There is currently no user defined gRPC middleware pipeline at all on either the Dapr API or the internal Dapr API. This means the behaviour of Dapr's HTTP and gRPC API are inconsistent. Should we be able to configure both gRPC pipelines? Do we need to rethink how this works? Can we abstract middleware over the protocol to unify the experience across HTTP and gRPC i.e. common middleware that take in a `request` object and a next `handler` and return a `response` and a protocol specific pipeline that handles the mapping to/from the abstract `request` and `response` to the native protocol implementation. Something like this would be needed for parts of [Proposal 2](#proposal-2).
- [ ] Add support for user defined gRPC middleware pipeline
- [ ] Consider unifying the middleware across protocols

# Review Summary
Dapr currently has a confusing (at least for me) set of middleware implementations and capabilities. I think we need to be very clear on what the overall design of middleware is and what sort of scenarios we want to enable. Is dapr going to try and compete with reverse proxies (envoy, traefik, nginx, haproxy, etc.) by supporting ingress style middleware?
in regards to authentication/authorization I think again we need to be clear on who is the expected client(s) of the Dapr API and therefore what sort of authentication/authorization mechanisms should be supported. I believe, the only client of the Dapr API should be the local app, and if this is the case, then mTLS + API tokens + JWTs should be sufficient for server side.

# Proposal(s)
Whilst trying to understand our current middleware's auth story and how Dapr can integrate with external policy engines such as Open Policy Agent (OPA) I came to the realization that policy could be applied more broadly across Dapr. I see _policy_ as a generalization of authorization and therefore put together a proposal for how we could integrate the concept of policy to enable authorization and also other features such as data validation. Below I have documented 2 proposals, the first is a grander plan to integrate Policy as a first class citizen in Dapr, the second is a simpler way to enable some additional authorization and policy integration.

## Proposal 1. Policy as a Building Block
> This would likely target post v1.0.0

### Policy Components
Define a new type of Policy component that can used by existing building blocks such as:
- [Middleware](#middleware-integration)
- [PubSub](#pubsub-integration)
- [State](#state-integration)
- [Dapr API](#policy-dapr-api)

The policy component defines a new interface `Check(r PolicyRequest) PolicyResponse`. The role of a Policy component is to make a policy decision based on a policy request. Where a policy is defined as a set of rules and a policy decision is used to determine whether or not a request satisfies said rules.

It is up to the policy component implementation how they satisfy this interface, some will handle the request locally, others can proxy out to external policy engines such as OPA.

Some potential policy component implementations could be:
- `component-contrib/policy/jwt/jwt.go` for simple JWT claims matching
- `component-contrib/policy/opa/opa.go` for using OPA as a Go library
- `component-contrib/policy/opa/opa_ext.go` for using OPA as an external service
- `component-contrib/policy/keycloak/keycloak.go` for using Keycloak proxy
- `component-contrib/policy/jsonschema/jsonchema.go` for JSON schema validation

### Middleware Integration
![authz](https://user-images.githubusercontent.com/7241190/83505752-246baf80-a4be-11ea-8edc-78607cf2e8d7.png)

We would need to add a new HTTP Policy middleware that could be configured to use a policy provider `middleware.http.policy` (gRPC too if supported).

You could define the Policy component itself using something like
```
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: http.jwt
spec:
  type: policy.jwt
  metadata:
  - name: aud
    value: "<allowed audience>"
  - name: iss
    value: "<allowed issuer>"
  - name: sub
    value: "<allowed subject>"
  - name: issuerURL
    value:  "<issuer URL>"
  - name: role
    value: "<role-1>;;<role-2>"
...
```
Then you could define the policy middleware component to leverage the policy component
```
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: middleware.http.policy.jwt
spec:
  type: middleware.http.policy
  metadata:
  - name: "engine"
  - value: "http.jwt" # Reference the Policy component name
```
Then when constructing a HTTP pipeline in configuration you could define it as such:
```
apiVersion: dapr.io/v1alpha1
kind: Configuration
metadata:
  name: pipeline
spec:
  httpPipeline:
    handlers:
    - name: middleware.http.oauth2
      type: middleware.http.oauth2
      selector:
        fields:
        - field: auth
          value: "path=invoke/pythonapp"
    - name: middleware.http.policy.jwt # Reference the Policy middleware name
      type: middleware.http.policy
      selector:
        fields:
        - field: auth
          value: "path=invoke/pythonapp"
```
This would add the oauth2 middleware to let users login and get an `access_token` for their Auth server. The `middleware.http.policy.jwt` would then construct a `PolicyRequest` from the HTTP request and context and send it to the `policy.http.jwt` policy implementation which would apply its configured rule processing and return a `PolicyResponse` indicating whether or not the request was compliant. The `middleware.http.policy.jwt` would then handle the decision according to its implementation design i.e. return 403 on deny. As the policy engine logic is encapsulated in a protocol agnostic package `component-contrib/policy/jwt/jwt.go` this could easily be used to create a gRPC equivalent. At the moment `middleware.http.policy.jwt` implementation would rely on fixed logic with some configuration on how to handle the policy decisions, but in the future could be expanded to use plugins, scripts or wasm to define custom policy decision handling logic.

With the policy middleware you could easily add simple policies for API constraints too such as maximum request size, body validation, etc. too.

### Pubsub Integration
![topic policy](https://user-images.githubusercontent.com/7241190/83505776-2f264480-a4be-11ea-877e-b3b647cd64e7.png)

As well as being invoked as part of the HTTP/gRPC pipeline, you could also integrate the policy engine check into the pubsub components. When a message is pulled from a topic, you could optionally check whether the message meets a policy before invoking the user apps handler. This would require some way to reference a Policy component from a pubsub component definition.

### State Integration
![state](https://user-images.githubusercontent.com/7241190/83505795-34838f00-a4be-11ea-9580-64399657caa2.png)

It is common boilerplate code to perform data validation and schema enforcement. We could allow users to express these rules in policy and have it enforced on the state API using a policy engine. This could be done via middleware applied to the state route or integrated into the state API directly as an optional call before invoking the state component.

### Policy Dapr API
![api](https://user-images.githubusercontent.com/7241190/83505806-38afac80-a4be-11ea-9e3e-09509bd9a152.png)

User code can also benefit from using the policy component. We could add policy to the Dapr API i.e. `/v1.0/policy/opa/` which they can use to enforce policy in their own app code. For instance, if the built in `middleware.http.policy` policy enforcement behaviour is not what you want, you could handle it yourself in your own app by using the policy API and having custom decision handling logic.

### Dapr components
Dapr components are currently evaluated by code, we could add a policy hook in to send the component to a policy engine to determine whether it meets a predefined policy. This way you would have greater control over how people use dapr in an organization. This would introduce a bootstrapping issue as a Policy would itself be a component but this could be handled in code.
Additionally, Dapr component have scopes - these are currently simple app_id name matches, but in theory you could put this behind a policy interface and allow people to configure more complicated rules about which sidecars can access which components such as, all dapr sidecars on windows nodes can access these components.
Granted that depending on the platform you may be able to perform this type of policy outside of Dapr i.e. on the Kubernetes object.

### Types
The types `PolicyRequest` and `PolicyResponse` would need to be defined in a common package such as  `component-contrib/policy/policy.go`. The `PolicyRequest` would be what the policy engines would be writing their policies against so would need to include enough information to cover most use cases.

We could split the `PolicyRequest` definition into multiple types i.e. `HTTPPolicyRequest`, `PubSubPolicyRequest`, etc. or we could use a single type with optional fields.
Things to consider including in the `PolicyRequest`:
- `dapr` 
  - `app_id`
  - `platform`
  - `component`
  - `version`
  - `etc.`
- `request` // only populated for request/response (invoke)
  - `http` // only populated on http request
     - `headers`
     - `method`
     - `path`
     - `host`
     - `scheme`
     - `size`
     - `body` (optional)
     - `protocol`
     - `query`
  - `grpc` // only populated on grpc request
    - `method`
    - `parameters`
    - `metadata`
  - `auth` // only populated when tokens present
     - `tokens` // contains pre-parsed tokens
- `pubsub` // only populated on pubsub message
   - `message`
   - `topic`
- `network`
  - `source`
    - `address`
    - `certificate` (optional)
  - `destination`
    - `address`
    - `certificate` (optional)
- `metadata`
   - `timestamp`

Things to consider including in the `PolicyResponse`:
- `status` // Did the request satisfy the policy
- `response`
     - `http` 
         - `request` // only populated when you want to modify the http request
           - `headers`
           - `body`
         - `response` // only populated when you want to modify the http response
           - `headers`
           - `body`
     - `grpc`
         - `error`
         - `metadata`

### Policy Lanauge
A simple JWT implementation could use some simple configuration on the Policy component to drive it's policy language. If it can support handling simple string and array matching it'll cover most use cases.

A more complex OPA policy could be expressed in [Rego](https://www.openpolicyagent.org/docs/latest/policy-language/) and supplement it with data as part of a bundle. The Policy configuration can instruct this implementaton where to load the bundles from etc. Or in the remote case, how to connect and authenticate to the remote OPA instance.

Using this model allows users to use existing tools and frameworks to define their policies and mix and match tools in a flexible way across Dapr.

## Proposal 1 Summary
Proposal 1 is a potential direction for longer term policy support in Dapr. It could be implemented incrementally by feature i.e. integrate with middleware first, add pubsub after. etc. Although I'm not expecting this to be implemented exactly as proposed, I wanted to do a detailed proposal to give the community and maintainers some ideas about how Dapr could bring value to policy.

# Proposal 2. Auth & Policy Middleware
Counter intuitively, proposal 2. is really a step towards proposal 1. and would be tackled first.
We can add some relatively small changes to the existing Bearer middleware to support JWT claims. We could also look to introduce an OPA HTTP middleware for more advanced request policies. However, this would limit the integration to just HTTP/gRPC pipeline.

I think to make middleware effective for service to service invocation we should add a capability to add middleware to the receiving side of the request. This would enable us to add authentication to the actual app we are trying to protect rather than the clients and leverage additional features such as rate limiting. Currently the internal gRPC server does not have a configurable middleware pipeline (beyond monitoring). As this is a gRPC server as well it only supports gRPC middleware - we could build on the work done in the [data exchange specification](https://github.com/dapr/dapr/issues/1409) to try and unify the middleware across the protocols as described in [Middleware](#grpc-middleware).
Another option might be to look to add a HTTP middleware pipeline on the receiver side and push the request through that before it's sent to the clients app but this feels a little less clean.
If we didn't want to have another configurable middleware pipeline we could just add explicit support for JWTs as a middleware on the existing internal gRPC API. This would allow us to just solve this particular issue of JWT support in service to service requests but wouldn't give us an extensible way to support other capabilities such as rate limiting.
![middleware_2](https://user-images.githubusercontent.com/7241190/83652825-f6b46280-a5b2-11ea-9370-2574cc381053.png)

### HTTP Bearer Middleware
In this design we simply add JWT claims handling to the existing [bearer middleware](https://github.com/dapr/components-contrib/blob/master/middleware/http/bearer/bearer_middleware.go). Putting the common token handling code in a package `/components-contrib/policy/jwt/jwt.go`.

### HTTP OPA Middleware
We add a new middleware that builds an OPA input from a HTTP request and uses the OPA library or service to get a policy decision. The middleware then uses the policy decision to invoke an appropriate action. This could be implement for both local OPA and remote OPA. All OPA related logic is captured in a common code package at `/components-contrib/policy/opa/opa.go`.

## Proposal 2 Summary
Proposal 2 is a short term solution that would unblock a number of scenarios for users and could be implemented fairly quickly. I believe the fixes outlined here should be strongly considered for pre v1.0.0 as these will satisfy common production requirement for users. Obviously they could write all this logic in their own apps if we didn't enable this feature but this wouldn't be a great experience.

- [ ] Add JWT claims handling to [Bearer HTTP Middleware](https://github.com/dapr/components-contrib/blob/master/middleware/http/bearer/bearer_middleware.go)
- [ ] Add OPA middleware
- [ ] Add middleware pipeline between Dapr and App

# References & Notes

## Open Policy Agent
Below is some context on Open Policy Agent (OPA) and how it could fit into this model.

OPA is a policy engine designed for cloud native environments that is currently being incubated in the CNCF. OPA supports a policy language called rego which allows you to express complex policy definitions that can be applied to input data. OPA also allows you to import supplementary data which can be used in your policy definitions. For instance, you could import a list of “owners” and use a rego policy to only allow them access to a resource. These assets (rego policies and data) can be provided to OPA as bundles (tarballs) from local disk or from a remote endpoint. OPA itself is typically deployed as a sidecar and can also support a centralized control-plane that includes bundles, auditing, status updates and asset discovery. OPA defines the interface to the control-plane services but does not itself provide an implementation.

The value proposition of OPA is that it decouples policy decisions from policy enforcement. Therefore, you can delegate complex policy parsing and rules to OPA whilst keeping the flexibility of data models and decision handling in your app.

### OPA Integration
There are 2 methods of integrating with OPA.
1.	Use OPA’s REST API
2.	Use OPA as a Go library

Given that both Dapr and OPA are written in Go we can leverage OPA as a library and as an external service via it's REST API. There is a perf trade off between these options; using the Go library will have reduced latency as there is not additional network hop, however, OPAs memory consumption for caching bundles etc. will now be embedded in Dapr. We should look to support both options to give the users a choice on which option they prefer in their scenario.

OPA does not mandate the structure of the input data that you provide for it to make a decision on. However, depending on the policy, it can return complex data that needs handling before a decision to allow or deny a request can be made.

## WIP OPA Middleware
I've started a quick test implementation of a HTTP middleware using OPA to enforce a policy on a HTTP request based on it’s method and path (https://github.com/jjcollinge/components-contrib/tree/jjcollinge/opa-mw). I’ll continue to play with this, move it closer to this proposal and revise as I learn more about OPA.

Would welcome early feedback and suggestions!

cc: @RichiCoder1

Related:
https://github.com/dapr/dapr/issues/1350
https://github.com/dapr/components-contrib/issues/189
https://github.com/dapr/components-contrib/issues/155

**Comment by RichiCoder1** [comment]
Looks solid! I wonder if there's any more Dapr specific metadata we can provide as inputs. Also might expand on what "response" in PolicyResponse means. Having using Envoy's ext_authz, sometimes you want to reject with a full response, sometimes you just wanna add a header to the downstream request, or a header to the _upstream_ response

**Comment by RichiCoder1** [comment]
For an MVP though, just being able to set a status and optionally override the response might be good enough to start with.

**Comment by benc-uk** [comment]
Proposal 1 is interesting it's the first case in Dapr that I've seen where components reference other components. 
I like that's it starting to become a higher level of set of abstractions meeting common application scenarios from the raw/generic building blocks Dapr currently has

I think we need to be careful with the naming of the policy component and the policy middleware component, at first glance it's not obvious what is doing the enforcement and rule checking. Maybe just me going word blind on the word "policy"

This could move a lot of tasks developers do in code into configuration, so is well aligned to what I see as the core benefits of Dapr. It's even a baby step towards a PaaS like experience 

Proposal 1 represents a great stepping back and looking at the wider problem. Proposal 2 is obviously more tactical

@jjcollinge BTW I looked at OPA support for validation of JSON and https://github.com/open-policy-agent/opa/issues/1449 it's an long time open request.
I think a better option is to probably have another policy engine component that implements JSON schema validation, rather than trying to bend OPA into doing that, e.g. a wrapper around something like https://github.com/xeipuuv/gojsonschema

**Comment by benc-uk** [comment]
The middleware selectors PR https://github.com/dapr/dapr/pull/1557 is obviously key here, regardless of which route is taken, the current "all traffic" approach is an impediment

**Comment by youngbupark** [comment]
@jjcollinge Thanks for the proposal.

@msfussell @orizohar @amanbha @yaron2 Please review this proposal. On the other hand, I am not sure that this is the right time to add this toward 1.0.

**Comment by RichiCoder1** [comment]
This issue got a serious update! I think starting with a middleware would be a great short-term (hopefully pre-or-at-1.0) step, and then expand it into a fuller policy framework.

**Comment by jjcollinge** [comment]
> This issue got a serious update! I think starting with a middleware would be a great short-term (hopefully pre-or-at-1.0) step, and then expand it into a fuller policy framework via P2.

@RichiCoder1 haha Yh, I think I got a little carried away! 🤔 definitely trying to position P2 as something tangible and P1 more as a vision. I would be interested in your use case with Gloo - is the current middleware pipeline sufficient (app to Dapr) or would you require the Dapr to app middleware proposed on P2 to protect your own APIs?

**Comment by RichiCoder1** [comment]
At least _right now_, P2 more than suits our case as we just want to apply policies to requests coming via the API Gateway. In the future we'd like to apply those same policies to inter-service communication too, but that'd also be covered (minus gRPC not currently having middleware)

**Comment by jjcollinge** [comment]
@RichiCoder1 ok cool, and do those policies have to be expressed in OPA or would simple JWT claims handling be sufficient? - just trying to prioritize.

**Comment by RichiCoder1** [comment]
I think simple JWT checks would help a lot of people, but for our case we already have OPA policy bundles we're using. Some of our checks require more context than just the claims too (HTTP context), so simple claims checks wouldn't be enough for us. (Hence why I'm game to help)

**Comment by jjcollinge** [comment]
Great, agree I think JWT will cover most. Would you want to have a crack at the OPA middleware yourself? Happy to help as I’m sure others are too. I can take a crack at updating the JWT middleware later this week.

**Comment by RichiCoder1** [comment]
Sure! I've still a bit of a Go novice, but I've been wanting to dig into both Go and Dapr, so I'll take the excuse :)

**Comment by yaron2** [comment]
@RichiCoder1 have you made progress on an OPA middleware?

**Comment by RichiCoder1** [comment]
@yaron2 Sadly haven't had the cycles. Still would likely tackle/assist but probably won't be till this weekend or next.

**Comment by RichiCoder1** [comment]
Update: Finally setting aside cycles to tackle this. Got some experience integrating OPA via other work

**Comment by mchmarny** [comment]
+1 on a lib to minimize any latency implications related to HTTP serde to start with and we evolve based on real-world usage from there. 

Overall, IMO the desired approach here is one that minimizes the user exposure to OPA implementation on Dapr side. While Rego is go-like (+/- for some), using the go library approach would allow us to potentially provide multiple ways of specifying the policies as the usage evolves. For example, initially by providing snippets or paths to the policy manifests, and perhaps eventually through alternative (more user-friendly?) ways, which would convert Dapr idiomatic user intent into an OPA policy on the backend.

**Comment by RichiCoder1** [comment]
That seems reasonable. For a first-pass MVP, add via Go SDK w/ support for rego snippets embedded in config? Just integrating OPA and defining the `input` will probably be a lot of work, so I'm fine with a tight scope. Then layer other features on top in priority, like bundles, discovery, and decision logging.

> perhaps eventually through alternative (more user-friendly?) ways, which would convert Dapr idiomatic user intent into an OPA policy on the backend.

That'd be neat. I forget where but I do remember seeing some RBAC/ABAC framework that took configs and compiled that down to rego.

**Comment by yaron2** [comment]
We also need to watch out for the increase in the binary size of Dapr if we go the SDK route.

**Comment by RichiCoder1** [comment]
Added the rough outline of a very MVP implementation of OPA Middleware here: https://github.com/dapr/components-contrib/pull/456.
First major outing into Go land too.

**Comment by yaron2** [comment]
> Added the rough outline of a very MVP implementation of OPA Middleware here: [dapr/components-contrib#456](https://github.com/dapr/components-contrib/pull/456).
> First major outing into Go land too.

Awesome! We'll review

**Comment by RichiCoder1** [comment]
The OPA Http Portion of this is integrated, but I realized I deviated from the design here somewhat in that the inputs and outputs are HTTP specific. @yaron2 I may make one more change before release to at least change the `package` and `input` to reflect @jjcollinge proposed structure with the HTTP input being `input.request.http` rather than just `input.request`.

**Comment by RichiCoder1** [comment]
That said, the next step after that would likely be extracting the OPA code into a reusable policy component as outlined here, getting bundle support working, and updating the middleware to be able to reference the policy component.

**Comment by yaron2** [comment]
> That said, the next step after that would likely be extracting the OPA code into a reusable policy component as outlined here, getting bundle support working, and updating the middleware to be able to reference the policy component.

That's currently out of scope. General policy will be revisited in the future.

**Comment by RichiCoder1** [comment]
Sounds good @yaron2. Just let me know how I can chip in!

**Comment by yaron2** [comment]
> The OPA Http Portion of this is integrated, but I realized I deviated from the design here somewhat in that the inputs and outputs are HTTP specific. @yaron2 I may make one more change before release to at least change the `package` and `input` to reflect @jjcollinge proposed structure with the HTTP input being `input.request.http` rather than just `input.request`.

The middleware itself is registered as an HTTP middleware, so I don't see changes needed here..

**Comment by RichiCoder1** [comment]
That's the conclusion I came to too after some though. If ya'll make the policy structure more generic, people would have to adjust their policies _but_ they'd likely have to do that anyway.

**Comment by msfussell** [comment]
@jjcollinge - After having JWT and OPA middleware in Dapr, what is now the delta from this proposal?

**Comment by jjcollinge** [comment]
@msfussell sorry for the delay in getting back to you. Adding the OPA and OAuth2 client credentials has definitely closed the gap on this issue. However, there are a few items that I still believe we need to at least discuss before 1.0.
- HTTP Middleware on the Callee side. You can then protect your API as the API host not as the API client.
- HTTP Middleware selectors for applying middleware to specific routes. (PR available)
- gRPC middleware?
These items may have been addressed but I haven't seen anything to confirm the decisions.

Cc: @mchmarny @yaron2
